actors/builtin/market/deal.go

```diff
@@ -12,6 +24,167 @@ import (
+	if !label.IsString() {
+		return "", xerrors.Errorf("label is not string")
+	}
+
+	return string(label.bs), nil
+}
+
+func (label DealLabel) ToBytes() ([]byte, error) {
+	if !label.IsBytes() {
+		return nil, xerrors.Errorf("label is not bytes")
+	}
+	return label.bs, nil
+}
+
+func (label DealLabel) Length() int {
+	return len(label.bs)
+}
+
+func (l DealLabel) Equals(o DealLabel) bool {
+	return bytes.Equal(l.bs, o.bs) && l.notString == o.notString
+}
+
+func (label *DealLabel) MarshalCBOR(w io.Writer) error {
+	scratch := make([]byte, 9)
+
+	// nil *DealLabel counts as EmptyLabel
+	// on chain structures should never have a pointer to a DealLabel but the case is included for completeness
+	if label == nil {
+		if err := cbg.WriteMajorTypeHeaderBuf(scratch, w, cbg.MajTextString, 0); err != nil {
+			return err
+		}
+		if _, err := io.WriteString(w, string("")); err != nil {
+			return err
+		}
+		return nil
+	} else if label.IsString() && label.IsBytes() {
+		return fmt.Errorf("dealLabel cannot have both string and bytes set")
+	} else if label.IsBytes() {
+		if len(label.bs) > cbg.ByteArrayMaxLen {
+			return xerrors.Errorf("labelBytes is too long to marshal (%d), max allowed (%d)", len(label.bs), cbg.ByteArrayMaxLen)
+		}
+
+		if err := cbg.WriteMajorTypeHeaderBuf(scratch, w, cbg.MajByteString, uint64(len(label.bs))); err != nil {
+			return err
+		}
+
+		if _, err := w.Write((label.bs)[:]); err != nil {
+			return err
+		}
+	} else { // label.IsString()
+		if len(label.bs) > cbg.MaxLength {
+			return xerrors.Errorf("labelString is too long to marshal (%d), max allowed (%d)", len(label.bs), cbg.MaxLength)
+		}
+
+		if err := cbg.WriteMajorTypeHeaderBuf(scratch, w, cbg.MajTextString, uint64(len(label.bs))); err != nil {
+			return err
+		}
+		if _, err := w.Write((label.bs)[:]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (label *DealLabel) UnmarshalCBOR(br io.Reader) error {
+	if label == nil {
+		return xerrors.Errorf("cannot unmarshal into nil pointer")
+	}
+
+	// reset fields
+	label.bs = nil
+
+	scratch := make([]byte, 8)
+
+	maj, length, err := cbg.CborReadHeaderBuf(br, scratch)
+	if err != nil {
+		return err
+	}
+
+	if maj == cbg.MajTextString {
+
+		if length > cbg.MaxLength {
+			return fmt.Errorf("label string was too long (%d), max allowed (%d)", length, cbg.MaxLength)
+		}
+
+		buf := make([]byte, length)
+		_, err = io.ReadAtLeast(br, buf, int(length))
+		if err != nil {
+			return err
+		}
+		s := string(buf)
+		if !utf8.ValidString(s) {
+			return fmt.Errorf("label string not valid utf8")
+		}
+		label.bs = buf
+		label.notString = false
+	} else if maj == cbg.MajByteString {
+
+		if length > cbg.ByteArrayMaxLen {
+			return fmt.Errorf("label bytes was too long (%d), max allowed (%d)", length, cbg.ByteArrayMaxLen)
+		}
+
+		bs := make([]uint8, length)
+		label.bs = bs
+		label.notString = true
+		if _, err := io.ReadFull(br, bs[:]); err != nil {
+			return err
```

Comment:
We should be able to unify these codepaths.

1. Make sure we have either a text string or a byte string.
2. Set "not string" if it's a byte string.
3. Check the length.
4. Read the value.
5. If a string, check if it's utf8.